zebra-state/src/service/finalized_state/zebra_db/block.rs

```diff
@@ -147,34 +147,28 @@ impl ZebraDb {
         }))
     }
 
-    /// Returns the Sapling
-    /// [`NoteCommitmentTree`](sapling::tree::NoteCommitmentTree) specified by a
-    /// hash or height, if it exists in the finalized `db`.
+    /// Returns the Sapling [`note commitment tree`](sapling::tree::NoteCommitmentTree) specified by
+    /// a hash or height, if it exists in the finalized state.
     #[allow(clippy::unwrap_in_result)]
-    pub fn sapling_tree(
+    pub fn sapling_tree_by_hash_or_height(
         &self,
         hash_or_height: HashOrHeight,
     ) -> Option<Arc<sapling::tree::NoteCommitmentTree>> {
         let height = hash_or_height.height_or_else(|hash| self.height(hash))?;
 
-        let sapling_tree_handle = self.db.cf_handle("sapling_note_commitment_tree").unwrap();
-
-        self.db.zs_get(&sapling_tree_handle, &height)
+        self.sapling_tree_by_height(&height)
     }
 
-    /// Returns the Orchard
-    /// [`NoteCommitmentTree`](orchard::tree::NoteCommitmentTree) specified by a
-    /// hash or height, if it exists in the finalized `db`.
+    /// Returns the Orchard [`note commitment tree`](orchard::tree::NoteCommitmentTree) specified by
+    /// a hash or height, if it exists in the finalized state.
     #[allow(clippy::unwrap_in_result)]
-    pub fn orchard_tree(
+    pub fn orchard_tree_by_hash_or_height(
         &self,
         hash_or_height: HashOrHeight,
     ) -> Option<Arc<orchard::tree::NoteCommitmentTree>> {
         let height = hash_or_height.height_or_else(|hash| self.height(hash))?;
 
-        let orchard_tree_handle = self.db.cf_handle("orchard_note_commitment_tree").unwrap();
-
-        self.db.zs_get(&orchard_tree_handle, &height)
+        self.orchard_tree_by_height(&height)
     }
```

Comment:
Thanks for catching this! I didn't notice this direct access to these column families when I made the compatibility fixes for this format change in `Zebra 1.0.0-rc.9`.

What this means is that the `z_gettreestate` RPC can return incorrect responses with missing trees if:
1. A state is upgraded by Zebra 1.2.0 (assuming this PR is merged after the next release)
2. A previous Zebra version is launched with that upgraded state

That's ok because backwards compatibility is best-effort, so we can just add a note to the changelog.

We could make this better by splitting this PR, merging these compatibility fixes in `1.1.0`, and then making the format changes in `1.2.0`. So we could tell users `1.1.0` is compatible, but older versions aren't.

Do you think it is worth the effort of doing this?

Reply:
I'd also like to get @arya2's feedback here, because if we split the PR, the compatibility fix becomes a release blocker.

Reply:
I think we should instead add config/CLI options behind a compilation feature for ignoring optional format upgrades and a subcommand for migrating down for users intending to switch back to an older version of Zebra.

Maybe Zebra could ask for confirmation before format upgrades too.

> Do you think it is worth the effort of doing this?

I think it's worth it this is to be a required format upgrade in v1.2.0.

Reply:
> > Do you think it is worth the effort of doing this?
> 
> I think it's worth it if this is to be a required format upgrade in v1.2.0.

Just checking we're talking about the same thing here: this PR currently contains a required format upgrade.

After it is merged, if blocks are finalized by a newer Zebra version, then read by an older Zebra version, some z_gettreestate responses will be incorrect.

So this is a question that I'd like to answer before the `1.1.0` release:

> > We could make this better by splitting this PR, merging these compatibility fixes in 1.1.0, and then making the format changes in 1.2.0. So we could tell users 1.1.0 is compatible, but older versions aren't.

I would be ok with just breaking compatibility of new states with old Zebra versions, and adding a note to the changelog, because running a newer state with an older Zebra version is an uncommon use case.

So do we want to split this PR and wait for a compatibility fix to merge before tagging the release, or just merge this PR and break compatibility for an uncommon use case?

(I asked Arya because he is the release manager and so he might need to delay the release.)

-----

This is a question that might be able to wait for a later release:

> I think we should instead add config/CLI options behind a compilation feature for ignoring optional format upgrades and a subcommand for migrating down for users intending to switch back to an older version of Zebra.
> 
> Maybe Zebra could ask for confirmation before format upgrades too.

All of our scheduled format upgrades are required to support the "spend before sync" RPCs and get reasonable performance. Since there are no optional format upgrades yet, can this design wait for the first release where there is an optional format upgrade?

At that time, we might want to consider how much benefit users get from compatibility with older Zebra versions, the extra implementation/test cost, ongoing maintenance cost / bug risk, and if an interactive prompt breaks existing non-interactive use cases.

Reply:
As an example, `zcashd`'s database capability flags cause significant ongoing user confusion. You can see some recent user questions about them in Discord.

So before we add optional state features or state upgrades, let's there's a compelling use case?

Reply:
> Just checking we're talking about the same thing here: this PR currently contains a required format upgrade.

I think it's worth the effort to split up this PR so the format upgrade is backwards compatible with v1.1.0 if this is to be a required format upgrade.

But I think we should make this an optional format upgrade with a CLI option and/or config field behind a compilation feature instead. Perhaps requiring user confirmation just before beginning the migration.

Reply:
> But I think we should make this an optional format upgrade with a CLI option and/or config field behind a compilation feature instead.

I'm a bit confused by this discussion, I feel like we're working towards different goals here. What goals are satisfied by making format upgrades optional and interactive? Are these goals important to users? Are they required for the 1.2.0 release?

We haven't scoped this work, so there aren't any tickets for it. It would significantly increase the complexity of our code, and our build and test matrixes, and ongoing maintenance. So I'm looking for a use case that justifies this extra effort.

> Perhaps requiring user confirmation just before beginning the migration.

Zebra currently doesn't ask for user input, so this would be a breaking change for all non-interactive use cases. I think we should wait for a major version release to break these common use cases.

-----

Here are the goals of requiring the two currently scheduled format upgrades:
* support "spend before sync" RPCs
* fast creation of the new database column family for the "z_getsubtreeindexes" RPC

They are important to users because they significantly increase light wallet sync time. They are required for the 1.2.0 or 1.3.0 release because zcashd already supports these RPC fields. So light wallet server operators and users are less likely to migrate to Zebra until they are done.

The user experience of requiring these changes is that most users won't notice. If users upgrade Zebra to 1.2.0 then downgrade to 1.0.1 or earlier, then some of the outputs of the "z_gettreestate" RPC might be wrong. But since we currently have 1 known user running a light wallet server, we can just tell them not to do that.

Reply:
I'm going to have to hand this over to @oxarbitrage and @mpguerra for a final decision, hopefully you can all discuss it at the engineering sync next week.

Reply:
> We haven't scoped this work, so there aren't any tickets for it.

I think it may be worth the effort to split this PR so v1.1.0 works with the new db format in that case.

Reply:
> > We haven't scoped this work, so there aren't any tickets for it.
> 
> I think it may be worth the effort to split this PR so v1.1.0 works with the new db format in that case.

Yep, I think that's the only decision we have to make for 1.1.0.

All the other decisions can wait for 1.2.0 or later.

Reply:
If possible, let's do the simplest and quickest thing for now to not block the next release and to give us time for a more in depth discussion later.

Reply:
I suggest we just merge this PR as is without splitting it.

Reasons:

- It is the simplest and quickest solution.
- The incompatibility impact is very low since someone would have to run a new Zebra version, which would remove some trees from the database, and then start an older version, which would end up returning `None` for heights for which the trees were removed in the `z_gettreestate` RPC.
- This behavior is not unexpected since, as Teor said, backward compatibility is best-effort.